src/storage/hummock_sdk/src/key.rs

```diff
@@ -575,39 +575,47 @@ impl UserKey<Vec<u8>> {
 #[derive(Clone, Copy, PartialEq, Eq, Hash, Default)]
 pub struct FullKey<T: AsRef<[u8]>> {
     pub user_key: UserKey<T>,
-    pub epoch: HummockEpoch,
+    pub epoch_with_gap: EpochWithGap,
 }
 
 impl<T: AsRef<[u8]>> Debug for FullKey<T> {
     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-        write!(f, "FullKey {{ {:?}, {} }}", self.user_key, self.epoch)
+        write!(
+            f,
+            "FullKey {{ {:?}, {} }}",
+            self.user_key,
+            self.epoch_with_gap.get_epoch()
+        )
     }
 }
 
 impl<T: AsRef<[u8]>> FullKey<T> {
-    pub fn new(table_id: TableId, table_key: TableKey<T>, epoch: HummockEpoch) -> Self {
+    pub fn new(table_id: TableId, table_key: TableKey<T>, epoch_with_gap: HummockEpoch) -> Self {
         Self {
             user_key: UserKey::new(table_id, table_key),
-            epoch,
+            epoch_with_gap: EpochWithGap::new(epoch_with_gap),
         }
     }
 
     pub fn from_user_key(user_key: UserKey<T>, epoch: HummockEpoch) -> Self {
-        Self { user_key, epoch }
+        Self {
+            user_key,
+            epoch_with_gap: EpochWithGap::new(epoch),
+        }
     }
 
     /// Pass the inner type of `table_key` to make the code less verbose.
-    pub fn for_test(table_id: TableId, table_key: T, epoch: HummockEpoch) -> Self {
+    pub fn for_test(table_id: TableId, table_key: T, epoch_with_gap: HummockEpoch) -> Self {
         Self {
             user_key: UserKey::for_test(table_id, table_key),
-            epoch,
+            epoch_with_gap: EpochWithGap::new(epoch_with_gap),
         }
     }
 
     /// Encode in to a buffer.
     pub fn encode_into(&self, buf: &mut impl BufMut) {
         self.user_key.encode_into(buf);
-        buf.put_u64(self.epoch);
+        buf.put_u64(self.epoch_with_gap.get_epoch());
```

Comment:
We may change the name of `get_epoch` to something like `as_u64`. The current `get_epoch` is misleading.